### DISCUSSION TITLE
Imagemagick php exec dont give any output

### DISCUSSION THREAD

**Discussion by PetarGajevic** [ImageMagick/ImageMagick#2829] [Help]
Hi, i am having trouble on my web server running commands, i just want to resize image for example. This is my code :

exec("/usr/local/bin/convert    /var/www/vhosts/urbanone.me/httpdocs/urbanone/public\site-images\U-one-16.jpg
-resize 1600x2500
/var/www/vhosts/urbanone.me/httpdocs/urbanone/public\site-images\U-one-16.jpg
" , $output);

So this works on my local windows machine, ofc with diffrent paths, but on my server its not working.
$output is empty " [ ] ". Can anyone help? Thanks.

**Comment by t246246** [comment]
Is your local Windows running PHP and you typed exact the above?
Suppose you just typed inside double quote in your command line on your PC and get desired result,

You cannot use backslash as a separator of path if the server is not Windows but Unix.
And even suppose the server is Windows, you need to use `\\` instead of `\`. (Check syntax of double quote in PHP manual.)

**Comment by snibgo** [comment]
See https://www.php.net/manual/en/function.exec.php

I suggest you retrieve all the text lines, not just the last line.

I also suggest you use return_var to get the return status, and that you then take appropriate error action.

**Comment by PetarGajevic** [comment]
Thank you for replies, so now when i do this:

exec("/usr/local/bin/convert  /var/www/vhosts/urbanone.me/httpdocs/urbanone/public/site-images/U-one-16.jpg
-resize 1600x2500
 /var/www/vhosts/urbanone.me/httpdocs/urbanone/public/site-images/U-one-16.jpg
" ,  $output,  $return_var  );

  dd($output , $return_var);

, the output is  ImageMagick version and list of commands and return_var is code 2.
I am not sure what return code 2 means?

**Reply by t246246** [reply]
I believe exit value from ImageMagick is 0 or 1.
But PHP's exec use shell to interpret command, so this is the return value from shell.
If your shell is bash, 2 means Misuse of shell builtins. See https://tldp.org/LDP/abs/html/exitcodes.html

Did you write the command in 1 line?

You said the first line of your output is ImageMaick version, this suggest some wrong element is passed to convert.

In usual case,

```
% php -r '$out=[]; exec("/usr/local/bin/convert logo.jpg -resize 100x100 logo.jpg", $out, $r); var_dump($out); print("$r\n");'
array(0) {
}
0
```
```
% php -r '$out=[]; exec("/usr/local/bin/convert -verbose logo.jpg -resize 100x100 logo.jpg", $out, $r); var_dump($out); print("$r\n");'
logo.jpg JPEG 100x75 100x75+0+0 8-bit sRGB 4195B 0.000u 0:00.001
array(1) {
  [0]=>
  string(64) "logo.jpg JPEG 100x75 100x75+0+0 8-bit sRGB 4195B 0.000u 0:00.001"
}
0
```

If you passed no argument to convert,
```
% php -r '$out=[]; exec("/usr/local/bin/convert", $out, $r); var_dump($out); print("$r\n");'
array(335) {
  [0]=>
  string(76) "Version: ImageMagick 7.0.10-35 Q16 x86_64 2020-11-01 https://imagemagick.org"
  [1]=>
  string(46) "Copyright: © 1999-2020 ImageMagick Studio LLC"
  [2]=>

...

[334]=>
  string(33) "'-' for standard input or output."
}
0
```

**Comment by snibgo** [comment]
>> exec("/usr/local/bin/convert /var/www/vhosts/urbanone.me/httpdocs/urbanone/public/site-images/U-one-16.jpg
-resize 1600x2500
/var/www/vhosts/urbanone.me/httpdocs/urbanone/public/site-images/U-one-16.jpg
" , $output, $return_var );

Is your command actually split into three lines? Is that allowed in PHP, or do you need a line-continuation character?

>> the output is ImageMagick version and list of commands 

That what you get if you run "convert" with no options at all, or just an input file.

**Comment by PetarGajevic** [comment]
Okay so now i used Process Component (https://symfony.com/doc/current/components/process.html) which i can include in my Laravel(PHP) project and  there is no too much diffrence between exec and process but i guess it is better for debugging. And error message it shows me " no decode delegate for this image format `JPG'" and when i run "convert -list configure"  i dont see delegates for jpg, png etc.. And it is command in one line. 
And i am sorry i am junior developer and new to ImageMagick.

**Reply by t246246** [reply]
How do you install imageMagick in your server?

Is your server Linux?
If so there was a problem in binary release recently.  See https://github.com/ImageMagick/ImageMagick/discussions/2774

Anyway I think you need to reinstall imageMagick with jpeg.
